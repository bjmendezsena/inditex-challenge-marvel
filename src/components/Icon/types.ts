export type IconSize = 'sm' | 'md' | 'lg' | 'xl';

export type IconProps = {
  size?: IconSize;
};
