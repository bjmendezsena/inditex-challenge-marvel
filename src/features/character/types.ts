import { URL, Stories, Thumbnail } from '@/types';
export interface IComic {
  id: number;
  digitalId: number;
  title: string;
  issueNumber: number;
  variantDescription: string;
  description: string;
  modified: string;
  isbn: string;
  upc: string;
  diamondCode: string;
  ean: string;
  issn: string;
  format: string;
  pageCount: number;
  textObjects: TextObject[];
  resourceURI: string;
  urls: URL[];
  series: Series;
  variants: any[];
  collections: any[];
  collectedIssues: any[];
  dates: DateElement[];
  prices: Price[];
  thumbnail: Thumbnail;
  images: Thumbnail[];
  creators: Creators;
  characters: Characters;
  stories: Stories;
  events: Characters;
}

export interface Characters {
  available: number;
  collectionURI: string;
  items: Series[];
  returned: number;
}

export interface Series {
  resourceURI: string;
  name: string;
}

export interface Creators {
  available: number;
  collectionURI: string;
  items: CreatorsItem[];
  returned: number;
}

export interface CreatorsItem {
  resourceURI: string;
  name: string;
  role: string;
}

export interface DateElement {
  type: string;
  date: string;
}

export interface Price {
  type: string;
  price: number;
}

export interface TextObject {
  type: string;
  language: string;
  text: string;
}
