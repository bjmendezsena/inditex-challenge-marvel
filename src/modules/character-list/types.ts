export type GetCharactersFilters = {
  nameStartsWith?: string;
  limit?: number;
};
