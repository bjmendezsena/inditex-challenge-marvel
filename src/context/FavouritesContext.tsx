import { createContext, PropsWithChildren, useState, useContext } from 'react';
import { Character } from '@/types';
import _ from 'lodash';
interface FavouritesContextType {
  favourites: Character[];
  addFavourite: (character: Character) => void;
  removeFavourite: (character: Character) => void;
  checkIsFavourite: (character: Character) => boolean;
  toggleFavourite: (character: Character) => void;
  setShowFavourites: (value: boolean) => void;
  showFavourites: boolean;
}

const FavouritesContext = createContext({} as FavouritesContextType);

export const FavouritesContextProvider = ({ children }: PropsWithChildren) => {
  const [favourites, setFavourites] = useState<Character[]>([]);

  const [showFavourites, setShowFavourites] = useState(false);

  const addFavourite = (character: Character) => {
    setFavourites([...favourites, character]);
  };

  const removeFavourite = (character: Character) => {
    setFavourites(
      favourites.filter(favourite => !_.isEqual(favourite, character))
    );
  };

  const checkIsFavourite = (character: Character) =>
    favourites.includes(character);

  const toggleFavourite = (character: Character) => {
    if (checkIsFavourite(character)) {
      return removeFavourite(character);
    }

    addFavourite(character);
  };

  return (
    <FavouritesContext.Provider
      value={{
        favourites,
        addFavourite,
        removeFavourite,
        checkIsFavourite,
        toggleFavourite,
        showFavourites,
        setShowFavourites,
      }}
    >
      {children}
    </FavouritesContext.Provider>
  );
};

export const useFavourites = () => useContext(FavouritesContext);
