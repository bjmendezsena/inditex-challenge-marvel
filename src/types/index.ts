export interface Character {
  id: number;
  name: string;
  description: string;
  modified: string;
  thumbnail: Thumbnail;
  resourceURI: string;
  comics: ComicsData;
  series: ComicsData;
  stories: Stories;
  events: ComicsData;
  urls: URL[];
}

export interface ComicsData {
  available: number;
  collectionURI: string;
  items: ComicsItem[];
  returned: number;
}

export interface ComicsItem {
  resourceURI: string;
  name: string;
}

export interface Stories {
  available: number;
  collectionURI: string;
  items: StoriesItem[];
  returned: number;
}

export interface StoriesItem {
  resourceURI: string;
  name: string;
  type: string;
}

export interface Thumbnail {
  path: string;
  extension: string;
}

export interface URL {
  type: string;
  url: string;
}

export interface Data<T> {
  offset: number;
  limit: number;
  total: number;
  count: number;
  results: T[];
}
