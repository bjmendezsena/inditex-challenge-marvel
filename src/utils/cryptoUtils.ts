import Crypto from "crypto-js";

export function createHashMd5(text: string): string {
  return Crypto.MD5(text).toString();
}
