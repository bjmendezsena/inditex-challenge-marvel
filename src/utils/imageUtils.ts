import { Thumbnail } from '@/types';

export function buildThumbnailImage(thumbnail: Thumbnail) {
  const { extension, path } = thumbnail;
  return `${path.replace('http', 'https')}.${extension}`;
}
