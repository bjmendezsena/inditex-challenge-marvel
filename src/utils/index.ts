export * from "./cryptoUtils";
export * from "./urls";
export * from "./format";
