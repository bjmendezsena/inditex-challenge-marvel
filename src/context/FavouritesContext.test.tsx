import userEvent from '@testing-library/user-event';
import { render, waitFor, getHtml } from '@/tests/utils';
import { getCharacterListMockResponse } from '@/tests';
import { Character } from '@/types';
import { FavouritesContextProvider, useFavourites } from './FavouritesContext';
type TestComponentProps = {
  testCharacter: Character;
};

describe('FavouritesContext', () => {
  const characters = getCharacterListMockResponse.results;
  const TestComponent = ({ testCharacter }: TestComponentProps) => {
    const { favourites, addFavourite, removeFavourite } = useFavourites();

    const handleAddFavourite = () => {
      addFavourite(testCharacter);
    };

    const handleRemoveFavourite = () => {
      removeFavourite(testCharacter);
    };

    return (
      <>
        <div>{favourites.length}</div>
        <div>{JSON.stringify(favourites)}</div>
        <button onClick={handleAddFavourite}>Add Favourite</button>
        <button onClick={handleRemoveFavourite}>Remove Favourite</button>
      </>
    );
  };
  const getComponent = (props: TestComponentProps) =>
    render(
      <FavouritesContextProvider>
        <TestComponent {...props} />
      </FavouritesContextProvider>
    );

  it('should render', () => {
    const { container } = getComponent({ testCharacter: characters[0] });
    expect(container).toMatchSnapshot();
  });

  it('should add and remove favourite', async () => {
    const character = characters[0];
    const { getByText, container } = getComponent({ testCharacter: character });

    await userEvent.click(getByText('Add Favourite'));

    await waitFor(() => expect(getByText('1')).toBeInTheDocument());

    const htmlString = getHtml(container);

    expect(htmlString).toContain(JSON.stringify(character));

    await userEvent.click(getByText('Remove Favourite'));

    await waitFor(() => expect(getByText('0')).toBeInTheDocument());
  });
});
