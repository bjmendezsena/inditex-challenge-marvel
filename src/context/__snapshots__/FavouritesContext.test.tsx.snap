// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavouritesContext > should render 1`] = `
<div>
  <div>
    0
  </div>
  <div>
    []
  </div>
  <button>
    Add Favourite
  </button>
  <button>
    Remove Favourite
  </button>
</div>
`;
